proposals/0015-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,362 @@
+
+To help to determine when the reward period completes, a new RPC API,
+`getNumOfRewardDistributionBlocks`,  will be added. This API will returns:
+
+   - `numOfRewardDistributionBlocks: u64`
+
+To query the reward received for individual account, user Apps will first make
+a query to `getNumOfRewardDistributionBlocks`, and wait for that many
+of blocks since the start of the epoch. Then, the App can call
+`getInflationReward` to query the rewards as before.
+
+`getNumOfRewardDistributionBlocks` will also be exposed via sysvar/syscall to
+support on chain Dapp program usage.
+
+
+### Restrict Stake Account Access
+
+To avoid the complexity and potential rewards in unexpected stake accounts,
+the stake program is disabled during the epoch reward period.
+
+Any transaction that invokes staking program during this period will fail with
+a new unique error code - `StakeProgramUnavailable`.
+
+That means all updates to stake accounts have to wait until the rewards
+distribution finishes.
+
+Because different stake accounts are receiving the rewards at different blocks,
+on-chain programs, which depends on the rewards of stakes accounts during the
+reward period, may get into partial epoch reward state. To prevent this from
+happening, loading stake accounts from on-chain programs during reward period
+will be disabled. However, reading the stake account through RPC will still be
+available.
+
+
+## Impact
+
+There are the two main impacts of the changes to stake accounts during the
+epoch rewarding phase.
+
+The first impact is that stake accounts will see their rewards being credited a
+few blocks later in the epoch than before.
+
+The second impact is that users will not be able to update their stakes during
+the epoch reward phases, and will have to wait until the end of the epoch
+reward period to make any changes.
+
+Nonetheless, the overall amount of time that the user must wait before
+receiving and updating their stake rewards should be roughly equivalent to what
+they are now experiencing on the current mainnet beta, since the prolonged
+first block time at the epoch boundary effectively obstructs the user's access
+to those stake accounts during that time.
+
+Another advantage with the new approach is that all non-staking transactions
+will continue to be processed, while those transactions are blocked on mainnet
+beta today.
+
+
+## Security Considerations
+
+While the proposed new approach does impact and modify various components of
+the validators, it does not alter the economics of the reward system.
+
+If all the changes are implemented correctly and tested fully, there are no
+security issues.
+
+
+## Backwards Compatibility
+
+This is a breaking change.  The new epoch calculation and distribution approach
+will not be compatible with the old approach.
+
+## Open Questions
+
+1. Should we just store the full rewards directly instead of the chunked
+   `EpochRewardReserve` SysVar accounts?
```

Comment:
I don't know if this can remain an open question for too long :sweat_smile:

I will still loudly say that one account makes this simpler to implement and track on-chain.

Reply:
I heard you. I will update the proposal to defer the multiple reserve sysvars to its own follow up SIMD.

Reply:
Sounds good, thanks!